### PR TITLE
ENH: add a new init method for k-means

### PR DESCRIPTION
**Pull request by vyasr**
This PR is aimed at completing PR #4119. That PR was mostly completely (aside from some random number issues that it appears we were willing to push to a separate PR), so I rebased onto master and finished up the outstanding comments that I could find. Let me know if there's anything additional I could do. We can close #4119 when this is merged.

### CODE REVIEW

File: scipy/cluster/tests/test_vq.py
Comment (rgommers) [inline scipy/cluster/tests/test_vq.py]: This causes a test failure:
```
warning_calls = (['cluster/tests/test_vq.py:258', 'optimize/optimize.py:814', 'sparse/__init__.py:248'], ['__init__.py:113', 'cluster/hierarchy.py:3547', 'cluster/vq.py:138', 'cluster/vq.py:579', 'cluster/vq.py:695', 'constants/codata.py:1196', ...])

    @pytest.mark.slow
    @pytest.mark.skipif(sys.version_info < (3, 4), reason="needs Python >= 3.4")
    def test_warning_calls_filters(warning_calls):
        bad_filters, bad_stacklevels = warning_calls
    
        # There is still one simplefilter occurrence in optimize.py that could be removed.
        bad_filters = [item for item in bad_filters
                       if 'optimize.py' not in item]
        # The filterwarnings call in sparse/__init__.py is needed.
        bad_filters = [item for item in bad_filters
                       if os.path.join('sparse', '__init__.py') not in item]
    
        if bad_filters:
            raise AssertionError(
                "warning ignore filter should not be used, instead, use\n"
                "scipy._lib._numpy_compat.suppress_warnings (in tests only);\n"
                "found in:\n    {}".format(
>                   "\n    ".join(bad_filters)))
E           AssertionError: warning ignore filter should not be used, instead, use
E           scipy._lib._numpy_compat.suppress_warnings (in tests only);
E           found in:
E               cluster/tests/test_vq.py:258

bad_filters = ['cluster/tests/test_vq.py:258']
bad_stacklevels = ['__init__.py:113', 'cluster/hierarchy.py:3547', 'cluster/vq.py:138', 'cluster/vq.py:579', 'cluster/vq.py:695', 'constants/codata.py:1196', ...]
warning_calls = (['cluster/tests/test_vq.py:258', 'optimize/optimize.py:814', 'sparse/__init__.py:248'], ['__init__.py:113', 'cluster/hierarchy.py:3547', 'cluster/vq.py:138', 'cluster/vq.py:579', 'cluster/vq.py:695', 'constants/codata.py:1196', ...])
```

You can fix it by using `suppress_warnings` from `scipy._lib._numpy_compat`. There's one other example of `suppress_warnings` usage in this file

**Comment by rgommers** [comment]
Pyflakes still has one nitpick as well:
```
scipy/cluster/vq.py:574:1: E305 expected 2 blank lines after class or function definition, found 1
1       E305 expected 2 blank lines after class or function definition, found 1
```

File: scipy/cluster/vq.py
Comment (rgommers) [inline scipy/cluster/vq.py]: This causes a rendering issue (from CI: `(WARNING/2) Enumerated list ends without a blank line; unexpected unindent.::`).

See section 13 of https://numpydoc.readthedocs.io/en/latest/format.html#sections for the correct formatting

File: scipy/cluster/vq.py
Comment (mdhaber) [inline scipy/cluster/vq.py]: I think this line needs to be deleted.

**Comment by rgommers** [comment]
Sphinx formatting can be fiddly. I think the last commit should work. If not, I usually align on the first letter of the content, so under `D` in `.. [1] D ...`

File: scipy/cluster/vq.py
Comment (vyasr) [reply scipy/cluster/vq.py]: Good catch thanks. I'll fix.

**Comment by vyasr** [comment]
This should be all set now.

File: scipy/cluster/vq.py
Comment (rgommers) [inline scipy/cluster/vq.py]: This is generally not the best idea - better to instantiate a new `RandomState` object, and allow for seeding. However, the way it's done here is consistent with the other two initialization methods that are already present. And there's an explicit example using `np.random.seed`. Hence I think we should merge this as is, and open a new issue for how random numbers are generated in all these methods.

Thoughts?

**Review by rgommers — APPROVED** [review]
All LGTM. I still need to go back and read the paper to verify the implementation is consistent with it. But code-wise this PR is good to go.

**Comment by VarIr** [comment]
Scikit-learn also has [k-means++](https://github.com/scikit-learn/scikit-learn/blob/f0ab589f/sklearn/cluster/k_means_.py#L43). Might be of interest for verification.

**Comment by rgommers** [comment]
> Scikit-learn also has k-means++. Might be of interest for verification.

Thanks. Indeed, I looked at that already. The implementations are slightly different, enough to make a direct comparison cumbersome. @VarIr if you're interested in this method, verification or use of this PR and confirmation that it works as expected for you would be very welcome.

File: scipy/cluster/vq.py
Comment (ev-br) [reply scipy/cluster/vq.py]: +1 for fixing the random number generator in one go

**Review by tylerjereddy — COMMENTED** [review]
@vyasr A fresh rebase / force push so the CI is up to date may be sensible.

Apart from that, there's an approving review from Ralf here & I'm inclined to merge sometime on Nov. 6th (tomorrow) if there are no objections.

**Comment by tylerjereddy** [comment]
Rebased / force-pushed as the CI run was a few months old

**Comment by vyasr** [comment]
Thanks for taking care of that, I didn't get to this yesterday. All good to go?

**Review by tylerjereddy — APPROVED** [review]
CI is all green so merging based on comments above. Might have been better if we had input on the algorithm behavior as noted above, but code is assessed as ok.

Probably helpful if you can add a release note of some sort [here](https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.2.0). Thanks @vyasr

**Comment by rgommers** [comment]
Great! Sorry it took so long @vyasr, thanks for finishing off this PR!

IIRC this looked good - I had a few things left to check, but fine to do that later on.

**Comment by rgommers** [comment]
> Probably helpful if you can add a release note of some sort here.

**Comment by rgommers** [comment]
> Probably helpful if you can add a release note of some sort here.

done

**Comment by rgommers** [comment]
> there's an approving review from Ralf here

Note that I have not been using approvals to indicate "this can be merged". Nor have others I believe. Approval often is just an approval of the changes in response to earlier review comments that have been addressed. Those earlier comments in many cases are only a partial review.

Limitation of the GitHub interface, but I believe most people are doing it like this.

File: scipy/cluster/vq.py
Comment (rgommers) [reply scipy/cluster/vq.py]: I opened gh-9450 for this

**Comment by tylerjereddy** [comment]
Alright, I'm not merging blindly in any case.

Some people certainly do use the approvals in that way -- indeed, some projects do use an approval threshold (i.e., 1 positive review) before the merge button is activated, which is a github option.

**Pull request merged**
